File: rust/src/solver/dijkstras.rs
use crate::solver::distances::{Distances};
use crate::maze::grid_maze::GridMaze;
use crate::maze::grid_node::GridNode;


/// finds the shortest path in the `maze`, beginning at `start` and finishing at `goal`
/// returns a `Distances` struct that only contains the positions of cells on the shortest
/// path
pub fn find_shortest_path(maze: &GridMaze, start: GridNode, goal: GridNode) -> Distances {
    // compute distances for all cells in the maze beginning at start Pos
    let maze_dist = maze.distances(&start);

    // start from the goal and work backwards towards start
    let mut cur_node = goal;

    // curr_path will only contain positions that are on the shortest path from goal to start
    let mut cur_path = Distances::new(start);
    // insert the current cell into curr_path
    cur_path.insert(cur_node, maze_dist[cur_node]);

    loop {
        if cur_node == start {
            break;
        }

        // get the positions of all neighbors to the current cell's position
        for neighbor_node in maze.get_links(&cur_node) {

            // if the neighbor's distance is less than the current cell's distance, insert
            // the neighbor cell into curr_path, and make that neighbor the current cell
            if maze_dist[neighbor_node] < maze_dist[cur_node] {
                cur_path.insert(neighbor_node, maze_dist[neighbor_node]);
                cur_node = neighbor_node;
                break;
            }
        }
    }

    cur_path
}

File: nodejs/src/solver/dijkstras.js

const Distances = require("../Distances.js");

/**
 * finds the shortest path in a maze using Dijkstra's algorithm,  beginning at `startCell`
 * and finishing at `goalCell`.
 *
 * @param startCell {Cell} the starting cell of a maze to begin at
 * @param goalCell {Cell} the goal (or end) cell that you want to reach
 * @returns {Distances} a `Distances` object that contains the cells on the shortest path
 */
function shortestPathToGoal(startCell, goalCell) {
    // generate distances from startCell to goalCell
    const mazeDistances = startCell.distances();

    // starting at goalCell, work backwards towards startCell.
    // curPath will hold all cells on the shortest path
    let curCell = goalCell;
    const curPath = new Distances(curCell);

    // put curCell on the curPath, this is where the path 'starts'
    curPath.put(curCell, mazeDistances.distanceOf(curCell));

    // keep going until we reach the startCell
    while (curCell !== startCell) {

        // get all cells linked to curCell
        for (let linkedCell of curCell.links()) {

            // if the linkedCell's distance is less than the current cell's distance, insert
            // the linkedCell into curPath, and make that linkedCell the current cell
            if (mazeDistances.distanceOf(linkedCell) < mazeDistances.distanceOf(curCell)) {
                curPath.put(linkedCell, mazeDistances.distanceOf(linkedCell));
                curCell = linkedCell;
                break;
            }
        }
    }

    return curPath;
}

exports.shortestPathToGoal = shortestPathToGoal;